### ISSUE TITLE
🐛 Bug Report: GitlabOrgDiscoveryEntityProvider only supports one group since 1.18.0

### ISSUE THREAD

**Issue by bakito** [backstage/backstage#20274]
### 📜 Description

In [v1.18.0](https://github.com/backstage/backstage/releases/tag/v1.18.0) the group turned required in the GitlabOrgDiscoveryEntityProvider.

As I use a self hosted gitlab instance I would like to be able to import all / most of the groups from my instance.
Unfortunately only one group is supported.

### 👍 Expected behavior

For self hosted instances It should be possible to import all, or at least multiple groups with subgroups into backstage.

### 👎 Actual Behavior with Screenshots

When not configuring a group the sync fails.

`
{"class":"GitlabOrgDiscoveryEntityProvider","level":"error","message":"GitlabOrgDiscoveryEntityProvider:gitlabBisdevdomCH refresh failed GraphQL errors: [{\"message\":\"Internal server error\"}]","plugin":"catalog","service":"backstage","stack":"Error: GraphQL errors: [{\"message\":\"Internal server error\"}]\n    at GitLabClient.getGroupMembers (/app/node_modules/@backstage/plugin-catalog-backend-module-gitlab/dist/cjs/GitlabDiscoveryEntityProvider-8ed2153b.cjs.js:197:15)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async GitlabOrgDiscoveryEntityProvider.refresh (/app/node_modules/@backstage/plugin-catalog-backend-module-gitlab/dist/index.cjs.js:307:26)\n    at async TaskWorker.fn (/app/node_modules/@backstage/plugin-catalog-backend-module-gitlab/dist/index.cjs.js:240:13)\n    at async TaskWorker.runOnce (/app/node_modules/@backstage/backend-tasks/dist/index.cjs.js:349:7)\n    at async /app/node_modules/@backstage/backend-tasks/dist/index.cjs.js:297:31","target":"GitlabOrgDiscoveryEntityProvider:gitlabBisdevdomCH","taskId":"GitlabOrgDiscoveryEntityProvider:gitlabBisdevdomCH:refresh","taskInstanceId":"4931fee3-2407-4fd2-9ce8-d88414aff514","type":"plugin"}
`

If I define a group, only this one and it's members are imported.

### 👟 Reproduction steps

configure the GitlabOrgDiscoveryEntityProvider as documented in https://backstage.io/docs/integrations/gitlab/org

### 📃 Provide the context for the Bug.

```
  providers:
   
    gitlab:
      my-instance:
        host: gitlab.my.com
        orgEnabled: true
        schedule: # optional; same options as in TaskScheduleDefinition
          # supports cron, ISO duration, "human duration" as used in code
          frequency: { minutes: 30 }
          # supports ISO duration, "human duration" as used in code
          timeout: { minutes: 3 }
```

### 🖥️ Your Environment

Backstage 1.18.3

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by jamieklassen** [comment]
Sounds like this behaviour was probably introduced by #18889 -- however it was only supposed to make the group parameter required for `gitlab.com` and self-hosted installations weren't supposed to be affected. I don't really know what's going on here and don't have the bandwidth to properly investigate, but I'll take a long shot -- @matteosilv you think this could be a similar type of problem to the one solved in #20145?

**Comment by matteosilv** [comment]
Hi guys. No @jamieklassen it's a different problem (not related to #20145). However, seems weird log, since it seems the instance is recognized as SaaS. As you can see from the code below it only looks at group members for gitlab.com:

    if (client.isSelfManaged()) {
      groups = paginated<GitLabGroup>(options => client.listGroups(options), {
        page: 1,
        per_page: 100,
      });

      users = paginated<GitLabUser>(options => client.listUsers(options), {
        page: 1,
        per_page: 100,
        active: true,
      });
    } else {
      groups = (await client.listDescendantGroups(this.config.group)).items;
      users = (
        await client.getGroupMembers(this.config.group.split('/')[0], [
          'DIRECT',
          'DESCENDANTS',
        ])
      ).items;
    }

**Comment by matteosilv** [comment]
I just tested in my local instance updated to 1.18 and I am able to parse all groups starting from the root.

Can you verify also the `integrations` field in the conf yaml?

```yaml
integrations:
  gitlab:
    - host: gitlab.my.com
      token: ${GITLAB_ACCESS_TOKEN}
      apiBaseUrl: https:/gitlab.my.com/api/v4
```

**Comment by bakito** [comment]
> I just tested in my local instance updated to 1.18 and I am able to parse all groups starting from the root.
> 
> Can you verify also the `integrations` field in the conf yaml?
> 
> ```yaml
> integrations:
>   gitlab:
>     - host: gitlab.my.com
>       token: ${GITLAB_ACCESS_TOKEN}
>       apiBaseUrl: https:/gitlab.my.com/api/v4
> ```

@matteosilv Let me check that.
But I'll be only able to provide the info next week, as I don't have access to the system right now.

**Comment by jboeijenga** [comment]
We found some cause of this issue.
The graphQL query that is returning this Internal server error is requesting for the state of the users in the group

```
query getGroupMembers(
  $group: ID!
  $relations: [GroupMemberRelation!]
  $endCursor: String
) {
  group(fullPath: $group) {
    groupMembers(
      first: 100
      relations: $relations
      after: $endCursor
    ) {
      nodes {
        user {
          id
          username
          commitEmail
          name
          state
          webUrl
          avatarUrl
        }
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
  }
}
```
In our case we had some users with the state 'banned'. For some reason this was giving issues in the query. For us it was easy to overcome by giving the users another state since it was only a small number of users.

**Comment by jboeijenga** [comment]
I came across this issue as well and I think I found the rootcause.

The graphQL query that is used to fetch the group users also requests the users state:

```
query getGroupMembers($group: ID!, $relations: [GroupMemberRelation!], $endCursor: String) {
  group(fullPath: $group) {
    groupMembers(first: 100, relations: $relations, after: $endCursor) {
      nodes {
        user {
          id
          username
          commitEmail
          name
          state
          webUrl
          avatarUrl
        }
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
  }
}

```

However this state enum does not contain the banned state.
<img width="489" alt="Screenshot 2023-10-05 at 09 49 36" src="https://github.com/backstage/backstage/assets/1516438/59aec1ad-7c5b-41dc-b7ab-eb3367dea388">

Therefore, when we run this query for a group that contains a banned member you will receive the `Internal server error`

I solved it by changing the banned state of the banner users to blocked and the query was running fine again.

I create an MR with gitlab to add the missing `banned` value to the UserStateEnum https://gitlab.com/gitlab-org/gitlab/-/merge_requests/133398

**Comment by levanvunam** [comment]
> I came across this issue as well and I think I found the rootcause.
> 
> The graphQL query that is used to fetch the group users also requests the users state:
> 
> ```
> query getGroupMembers($group: ID!, $relations: [GroupMemberRelation!], $endCursor: String) {
>   group(fullPath: $group) {
>     groupMembers(first: 100, relations: $relations, after: $endCursor) {
>       nodes {
>         user {
>           id
>           username
>           commitEmail
>           name
>           state
>           webUrl
>           avatarUrl
>         }
>       }
>       pageInfo {
>         endCursor
>         hasNextPage
>       }
>     }
>   }
> }
> ```
> 
> However this state enum does not contain the banned state. <img alt="Screenshot 2023-10-05 at 09 49 36" width="489" src="https://user-images.githubusercontent.com/1516438/272822270-59aec1ad-7c5b-41dc-b7ab-eb3367dea388.png">
> 
> Therefore, when we run this query for a group that contains a banned member you will receive the `Internal server error`
> 
> I solved it by changing the banned state of the banner users to blocked and the query was running fine again.
> 
> I create an MR with gitlab to add the missing `banned` value to the UserStateEnum https://gitlab.com/gitlab-org/gitlab/-/merge_requests/133398

For my case seems `commitEmail` is not available in the GraphQL schema, only `publicEmail` available.

| 'commitEmail' doesn't exist on type 'UserCore'

My company uses a self host GitLab version: 15.7.7

Do you know if I need to enable something in GitLab to have the property or have to upgrade GitLab to have it?

Thank you

**Comment by bakito** [comment]
We do not have users in state 'banned'.

**Comment by jboeijenga** [comment]
@bakito It seems more statuses are missing in that enum, also users for who activation is pending or blocked by the system will cause the graph endpoint to throw this error, see my MR at Gitlab. 

To verify if this is the same root-cause you could copy the graph query from my earlier comment and go to the graphql-explorer at `https://gitlab.example.com/-/graphql-explorer` (replace `gitlab.example.com` with your instance's URL since you use Gitlab self-hosted). You can then find the group that is causing this error, remove the state field and validate if result are given back.

**Comment by bakito** [comment]
@jboeijenga how should I find the faulty group, we have > 200 groups, I can not query each single one !

**Comment by jboeijenga** [comment]
Sure, that would be a waste of time :). 
What I did is add a console.log to /node_modules/@backstage/plugin-catalog-backend-module-gitlab/dist/cjs/GitlabDiscoveryEntityProvider-8ed2153b.cjs.js line 149 to log every group the query is done for. This way it's easy to determine which group is causing issues. You can then perform the graphQL query to see what property is causing the issue and use cursors to determine the exact user that is causing this. good luck!

**Comment by bakito** [comment]
@jboeijenga I found one user with status `LDAP Blocked` which can not be loaded.

**Comment by bakito** [comment]
When checking gitlab logs I found the following error
```
Types::UserStateEnum::UnresolvedValueError (`User.state` returned `"ldap_blocked"` at `group.groupMembers.nodes.6.user.state`, but this isn't a valid value for `UserState`. Update the field or resolver to return one of `UserState`'s values instead.):
```

So it seems to be an issue with gitlab.

**Comment by pushit-tech** [comment]
Hello Team,

We are facing an issue regarding user retrieval, for not self hosted gitlab instance.

While I can fetch the groups successfully, not all users from my organization are being retrieved.

I'm curious if this concern aligns with the present issue, or if it necessitates the opening of a new one.

Upon examining the code, I observed that the switch from `getGroupMembers` to `listSaaSUsers` in [this commit](https://github.com/backstage/backstage/commit/753b35afb56268a2080875159f65ca1f449e51a2) led to the utilization of the REST API instead of GraphQL. 

By default, the REST API [only returns direct members and excludes inherited ones](https://docs.gitlab.com/ee/api/members.html#list-all-members-of-a-group-or-project), which may result in the query only fetching the users of the first group identified in this line: 

```javascript
const rootGroup = this.config.group.split('/')[0];
```

I found that we could continue using `listSaaSUsers` which calls `/groups/${encodeURIComponent(groupPath)}/members`, but by: 
- appending `\all` at the end 
- as `/groups/${encodeURIComponent(groupPath)}/members/all` 
- in accordance with the [GitLab documentation](https://docs.gitlab.com/ee/api/members.html#list-all-members-of-a-group-or-project), we can retrieve all users, including inherited members.

This adjustment could resolve the issue we're encountering, where only the users of the first group are being retrieved.
